### ISSUE TITLE
Network weights are not updating?!

### ISSUE THREAD

**Issue by kristoffzoghbi** [tiny-dnn/tiny-dnn#606]
I built a convolutional neural network with a structure similar to the network used in the Cifar10 example: 3 convolutional layers, 3 max pooling layers and 2 fully connected layers in the end.

After each epoch, I'm testing the accuracy of the network `net.test(t_images, t_labels).print_detail(cout)` on 1000 test items but the accuracy remains exactly the same! I tried changing the learning rate and other hyperparameters but to no avail. 

Also, I tried switching from a classification task to a regression task (`net.train()` to `net.fit()`) and the loss (`net.get_loss()`) is also exactly the same for the same items. 

Any idea why the weights are not updating?

**Comment by Randl** [comment]
Please, provide the code.

**Comment by kristoffzoghbi** [comment]
@Randl My code is spread over multiple classes but I've simplified it into this one class: http://pastebin.com/gPtgaWu3

**Comment by edgarriba** [comment]
@kristoffzoghbi how do you know that your weighs are not updated?

**Comment by kristoffzoghbi** [comment]
@edgarriba when I evaluate or test the network using the test dataset I always get the same accuracy or loss, even after many epochs of training. Even more, when I check the predicted value for a certain item from the dataset, it's always the same! I figured it must be that the weights aren't updating. What do you think?

**Comment by pliptor** [comment]
Maybe I'm reading your network wrong but it seems you have only one output at the last layer. Are you looking into a binary classification using Cifar10 data? One way you could try to troubleshoot is to dump the state of each layer as an image. tiny-dnn has such an interface.

**Comment by kristoffzoghbi** [comment]
@pliptor I'm using a network VERY similar to the network used in Cifar10 example but I'm not using Cifar10 data! I modified the network's last output layer to have 1 output to be used with a regression task.

I've tried to look into the layers' states by dumping weights and saving layers' output as images but couldn't find anything.

**Comment by Randl** [comment]
> I've tried to look into the layers' states by dumping weights and saving layers' output as images but couldn't find anything.

By found nothing you meant weights are unchanged? Did you try to increase learning rate?

**Comment by kristoffzoghbi** [comment]
@Randl I meant I got the same output every time. Same image, nothing changed. I also tried increasing the learning rate from 0.01 to 0.9 (using different rates in between).

**Comment by pliptor** [comment]
@kristoffzoghbi What type of data are you feeding as samples to the network? Do you normalize the input data before they are applied to the network? If they are not normalized, it could potentially get everything stuck.

**Comment by kristoffzoghbi** [comment]
@pliptor no I'm not normalizing the data. I should normalize since it's a regression task instead of classification? 
How can I normalize the `vec_t` input images?

**Comment by pliptor** [comment]
@kristoffzoghbi  It's kind of hard to tell without seeing the whole setup. I was just wondering why you were seeing no changes in the network and I thought one potential cause would be saturation, regardless if it was regression, classification, or any other system. Most systems have a dynamic range of operation.

The cifar10 example code uses this to parse the input data
[cifal10 parser](https://github.com/tiny-dnn/tiny-dnn/blob/master/tiny_dnn/io/cifar10_parser.h). 
Maybe you are already doing something similar since the prototype in your code above has a similar format.

**Comment by kristoffzoghbi** [comment]
@pliptor I've fixed the data encoding and **modified the activation function of the output layer from `softmax` to `activation::identity`** and it seems to work! Since it's a regression task the output has to be linear.

I will close this issue for now.

**Comment by pliptor** [comment]
@kristoffzoghbi I'm glad something worked! I haven't done regression tests yet but I see what you did
is recommended by others too.

https://deeplearning4j.org/linear-regression

 It seems then the saturation conjecture was correct. It was just happening at the output and not the input. (BTW, it reminded me that I do get better classification performance in some of my networks using identity activation is some layers).
